verification/certifyKey_test.go

```diff
@@ -124,15 +423,210 @@ func testCertifyKey(d TestDPEInstance, t *testing.T) {
+		foundError = true
+	}
+	t.Logf("Root certificate Signature validation passed %s\n", root.Subject.String())
+
+	if foundError {
+		// TODO: Fail the test when the certificate is expected to pass
+		t.Logf("Found certificates with mismatching signature")
+	}
+}
+
+// Build certificate chain and calls to validateSignature on each chain.
+func validateCertChain(t *testing.T, certChain []*x509.Certificate, leafCert *x509.Certificate) {
+	t.Helper()
+	if leafCert != nil {
+		// Remove unhandled critical extensions reported by x509 but defined in spec
+		t.Log("Validating leaf certificate chain...")
+		t.Log("Checking for unhandled critical certificate extensions unknown to DPE certificates profile spec...")
+		removeTcgDiceCriticalExtensions(t, []*x509.Certificate{leafCert})
+
+		// Build verify options
+		opts := buildVerifyOptions(certChain)
+
+		// Certificate chain validation for leaf
+		chains, err := leafCert.Verify(opts)
+		if err != nil {
+			// TODO: Fail the test with Errorf here once we expect it to pass.
+			t.Logf("Error in certificate chain %s: ", err.Error())
+		}
+
+		// Log certificate chains linked to leaf
+		for _, chain := range chains {
+			for i, cert := range chain {
+				t.Logf("%d %s", i, (*cert).Subject)
+			}
+
+			// Validate signature of all certificate chains
+			validateSignature(t, chain)
+		}
+
+		// This indicates that signature validation found no errors in the DPE leaf cert chain
+		t.Logf("DPE leaf certificate chain validation is done")
+
+	} else {
+		// Remove unhandled critical extensions reported by x509 but defined in spec
+		t.Log("Validating intermediate certificates chains...")
+		t.Log("Checking for unhandled critical certificate extensions unknown to DPE certificates profile spec...")
+		removeTcgDiceCriticalExtensions(t, certChain)
+
+		// Build verify options
+		opts := buildVerifyOptions(certChain)
+
+		// Certificate chain validation for each intermediate certificate
+		for _, cert := range certChain {
+			chains, err := cert.Verify(opts)
+			if err != nil {
+				// TODO: Fail the test with Errorf here once we expect it to pass.
+				t.Logf("Error in Certificate Chain of %s: %s", cert.Subject, err.Error())
+			}
+
+			// Log certificate chains linked to each cetificate in chain
+			for _, chain := range chains {
+				for i, cert := range chain {
+					t.Logf("%d %s", i, (*cert).Subject)
+				}
+				// Validate signature of all certificate chains
+				validateSignature(t, chain)
+			}
+		}
+
+		// This indicates that signature validation found no errors each cert
+		// chain of intermediate certificates
+		t.Logf("Intermediate certificates chain validation is done")
+	}
+}
+
+func buildVerifyOptions(certChain []*x509.Certificate) x509.VerifyOptions {
+	roots := x509.NewCertPool()
+	intermediates := x509.NewCertPool()
+
+	for _, cert := range certChain {
+		if cert.Subject.String() == cert.Issuer.String() {
```

Comment:
The root should be the highest cert in the chain. In DPE, the root is the certificate which occurs first in the GetCertificateChain response. All the other certs returned are intermediates.

Reply:
Ok, 
Added an additional check if cert in 0th index is a root cert.